mapogcapi.cpp

```diff
@@ -0,0 +1,1117 @@
+  try {
+    json geometry = getFeatureGeometry(shape, geometry_precision);
+    if(!geometry.is_null()) feature["geometry"] = geometry;
+  } catch (const std::runtime_error &e) {
+    throw std::runtime_error("Error fetching geometry.");
+  }
+
+  return feature;
+}
+
+static json getLink(hashTableObj *metadata, std::string name)
+{
+  json link;
+
+  const char *href = msOWSLookupMetadata(metadata, "A", (name + "_href").c_str());
+  if(!href) throw std::runtime_error("Missing required link href property.");
+
+  const char *title = msOWSLookupMetadata(metadata, "A", (name + "_title").c_str());
+  const char *type = msOWSLookupMetadata(metadata, "A", (name + "_type").c_str());
+
+  link = {
+    { "href", href },
+    { "title", title?title:href },
+    { "type", type?type:"text/html" }
+  };
+
+  return link;
+}
+
+static json getCollection(mapObj *map, layerObj *layer, int format)
+{
+  json collection; // empty (null)
+  rectObj bbox;
+
+  if(!map || !layer) return collection;
+
+  if(!msOWSRequestIsEnabled(map, layer, "AO", "OGCAPI", MS_FALSE) || !msWFSIsLayerSupported(layer)) return collection;
+
+  // initialize some things
+  std::string api_root = getApiRootUrl(map);
+
+  if(msOWSGetLayerExtent(map, layer, "AOF", &bbox) == MS_SUCCESS) {
+    if (layer->projection.numargs > 0)
+      msOWSProjectToWGS84(&layer->projection, &bbox);
+    else
+      msOWSProjectToWGS84(&map->projection, &bbox);
+  } else {
+    throw std::runtime_error("Unable to get collection bounding box."); // might be too harsh since extent is optional
+  }
+
+  const char *description = msOWSLookupMetadata(&(layer->metadata), "A", "description");
+  if(!description) description = msOWSLookupMetadata(&(layer->metadata), "OF", "abstract"); // fallback on abstract
+
+  const char *title = msOWSLookupMetadata(&(layer->metadata), "AOF", "title");
+
+  const char *id = layer->name;
+  char *id_encoded = msEncodeUrl(id); // free after use
+
+  // build collection object
+  collection = {
+    { "id", id },
+    { "description", description?description:"" },
+    { "title", title?title:"" },
+    { "extent", {
+	{ "spatial", {
+	    { "bbox", {{ bbox.minx, bbox.miny, bbox.maxx, bbox.maxy }}},
+	    { "crs", "http://www.opengis.net/def/crs/OGC/1.3/CRS84" }
+	  }
+	}
+      }
+    },
+    { "links", {
+        {
+          { "rel", format==OGCAPI_FORMAT_JSON?"self":"alternate" },
+          { "type", OGCAPI_MIMETYPE_JSON },
+          { "title", "This collection as JSON" },
+          { "href", api_root + "/collections/" + std::string(id_encoded) + "?f=json" }
+        },{
+          { "rel", format==OGCAPI_FORMAT_HTML?"self":"alternate" },
+          { "type", OGCAPI_MIMETYPE_HTML },
+          { "title", "This collection as HTML" },
+          { "href", api_root + "/collections/" + std::string(id_encoded) + "?f=html" }
+        },{
+          { "rel", "items" },
+          { "type", OGCAPI_MIMETYPE_JSON },
+          { "title", "Items for his collection as JSON" },
```

Comment:
minor spelling mistake, "Items for this collection as JSON"

Reply:
Fixed!